File: Services/pkg/user_role/repository.go
package user_role

import (
	"context"

	"github.com/ChangerzaryX1602/SkillSync/pkg/domain"
	"github.com/ChangerzaryX1602/SkillSync/pkg/models"
	txcontext "github.com/ChangerzaryX1602/SkillSync/pkg/tx_context"
	"github.com/ChangerzaryX1602/SkillSync/pkg/utils"

	"github.com/gofiber/fiber/v2"
	helpers "github.com/zercle/gofiber-helpers"
	"gorm.io/gorm"
)

type userRoleRepository struct {
	resources models.Resources
	cache     *userRoleCache
}

func NewUserRoleRepository(resources models.Resources) domain.UserRoleRepository {
	return &userRoleRepository{resources: resources, cache: newUserRoleCache(resources.RedisStorage)}
}

func (r *userRoleRepository) Migrate() error {
	return r.resources.MainDbConn.AutoMigrate(&models.UserRole{})
}

func (r *userRoleRepository) CreateUserRole(ctx context.Context, userRole models.UserRole) *helpers.ResponseError {
	db := r.getDB(ctx)
	if err := db.WithContext(ctx).Create(&userRole).Error; err != nil {
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	r.cache.InvalidateByUserID(userRole.UserID)
	return nil
}

func (r *userRoleRepository) GetUserRole(ctx context.Context, id uint) (*models.UserRole, *helpers.ResponseError) {
	db := r.getDB(ctx)
	var userRole models.UserRole
	if err := db.WithContext(ctx).Where("id = ?", id).First(&userRole).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return nil, &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return &userRole, nil
}

func (r *userRoleRepository) GetUserRoles(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.UserRole, *models.Pagination, *models.Search, *helpers.ResponseError) {
	db := r.getDB(ctx)
	var userRoles []models.UserRole
	db = db.WithContext(ctx).Model(&models.UserRole{})
	db = utils.ApplySearch(ctx, r.resources.FastHTTPClient, db, search, false)
	db = utils.ApplyPagination(db, &pagination, models.UserRole{})
	if err := db.Find(&userRoles).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil, nil, &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return nil, nil, nil, &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	return userRoles, &pagination, &search, nil
}

func (r *userRoleRepository) GetUserRolesByUserID(ctx context.Context, userID uint) ([]models.UserRole, *helpers.ResponseError) {
	if userRoles, found := r.cache.GetByUserID(userID); found {
		return userRoles, nil
	}
	db := r.getDB(ctx)
	var userRoles []models.UserRole
	if err := db.WithContext(ctx).Where("user_id = ?", userID).Find(&userRoles).Error; err != nil {
		return nil, &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	r.cache.SetByUserID(userID, userRoles)
	return userRoles, nil
}

func (r *userRoleRepository) UpdateUserRole(ctx context.Context, id uint, userRole models.UserRole) *helpers.ResponseError {
	db := r.getDB(ctx)
	if err := db.WithContext(ctx).Where("id = ?", id).First(&models.UserRole{}).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	if err := db.WithContext(ctx).Model(&models.UserRole{}).Where("id = ?", id).Updates(userRole).Error; err != nil {
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	r.cache.InvalidateByUserID(userRole.UserID)
	return nil
}

func (r *userRoleRepository) DeleteUserRole(ctx context.Context, id uint) *helpers.ResponseError {
	db := r.getDB(ctx)
	var userRole models.UserRole
	if err := db.WithContext(ctx).Where("id = ?", id).First(&userRole).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return &helpers.ResponseError{
				Code:    fiber.StatusNotFound,
				Source:  helpers.WhereAmI(),
				Title:   "Not Found",
				Message: err.Error(),
			}
		}
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	if err := db.WithContext(ctx).Delete(&models.UserRole{}, id).Error; err != nil {
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	r.cache.InvalidateByUserID(userRole.UserID)
	return nil
}

func (r *userRoleRepository) DeleteUserRolesByUserID(ctx context.Context, userID uint) *helpers.ResponseError {
	db := r.getDB(ctx)
	if err := db.WithContext(ctx).Where("user_id = ?", userID).Delete(&models.UserRole{}).Error; err != nil {
		return &helpers.ResponseError{
			Code:    fiber.StatusInternalServerError,
			Source:  helpers.WhereAmI(),
			Title:   "Database Error",
			Message: err.Error(),
		}
	}
	r.cache.InvalidateByUserID(userID)
	return nil
}

func (r *userRoleRepository) getDB(ctx context.Context) *gorm.DB {
	if tx := txcontext.GetTxFromContext(ctx); tx != nil {
		return tx
	}
	return r.resources.MainDbConn
}

File: hono-clean-architecture/pkg/user_role/repository.ts
import { NodePgDatabase } from "drizzle-orm/node-postgres";
import { eq, and, sql, count } from "drizzle-orm";
import { userRoles, UserRole, Pagination, Search, Resources } from "../models";
import { UserRoleRepository as IUserRoleRepository } from "../domain";
import { ResponseError, createError, whereAmI } from "../../internal/infrastructure/custom_error";
import { getOffset, getLimit } from "../utils";
import { UserRoleCache, newUserRoleCache } from "./cache";

export class UserRoleRepository implements IUserRoleRepository {
  private db: NodePgDatabase;
  private cache: UserRoleCache;

  constructor(resources: Resources) {
    this.db = resources.mainDbConn;
    this.cache = newUserRoleCache(resources.redisStorage);
  }

  async migrate(): Promise<void> {
    console.log("UserRole table migration handled by Drizzle Kit");
  }

  async createUserRole(data: { userId: number; roleId: number }): Promise<ResponseError | null> {
    const existing = await this.db
      .select()
      .from(userRoles)
      .where(and(eq(userRoles.userId, data.userId), eq(userRoles.roleId, data.roleId)))
      .limit(1);

    if (existing.length > 0) {
      return createError(409, whereAmI(), "Conflict", "UserRole already exists.");
    }

    try {
      await this.db.insert(userRoles).values({
        userId: data.userId,
        roleId: data.roleId,
      });

      await this.cache.invalidateAllLists();
      return null;
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return createError(500, whereAmI(), "Database Error", message);
    }
  }

  async getUserRole(
    id: number
  ): Promise<{ userRole: UserRole | null; error: ResponseError | null }> {
    const cached = await this.cache.get(id);
    if (cached) {
      return { userRole: cached, error: null };
    }

    try {
      const result = await this.db.select().from(userRoles).where(eq(userRoles.id, id)).limit(1);

      if (result.length === 0) {
        return {
          userRole: null,
          error: createError(404, whereAmI(), "Not Found", "UserRole not found"),
        };
      }

      await this.cache.set(result[0]);
      return { userRole: result[0], error: null };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return {
        userRole: null,
        error: createError(500, whereAmI(), "Database Error", message),
      };
    }
  }

  async getUserRoles(
    pagination: Pagination,
    search: Search
  ): Promise<{
    userRoles: UserRole[];
    pagination: Pagination;
    search: Search;
    error: ResponseError | null;
  }> {
    try {
      const countResult = await this.db.select({ count: count() }).from(userRoles);
      const total = Number(countResult[0]?.count ?? 0);

      let query = this.db.select().from(userRoles);

      if (pagination.orderBy) {
        const orderDirection = pagination.order === "desc" ? sql`DESC` : sql`ASC`;
        query = query.orderBy(
          sql`${sql.identifier(pagination.orderBy)} ${orderDirection}`
        ) as typeof query;
      }

      const result = await query.limit(getLimit(pagination)).offset(getOffset(pagination));

      return {
        userRoles: result,
        pagination: { ...pagination, total },
        search,
        error: null,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return {
        userRoles: [],
        pagination,
        search,
        error: createError(500, whereAmI(), "Database Error", message),
      };
    }
  }

  async deleteUserRole(id: number): Promise<ResponseError | null> {
    try {
      const userRole = await this.db.select().from(userRoles).where(eq(userRoles.id, id)).limit(1);

      if (userRole.length === 0) {
        return createError(404, whereAmI(), "Not Found", "UserRole not found");
      }

      await this.db.delete(userRoles).where(eq(userRoles.id, id));
      await this.cache.invalidate(id);
      await this.cache.invalidateAllLists();

      return null;
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return createError(500, whereAmI(), "Database Error", message);
    }
  }

  async getUserRolesByUserId(
    userId: number
  ): Promise<{ userRoles: UserRole[]; error: ResponseError | null }> {
    try {
      const result = await this.db.select().from(userRoles).where(eq(userRoles.userId, userId));

      return { userRoles: result, error: null };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Database error";
      return {
        userRoles: [],
        error: createError(500, whereAmI(), "Database Error", message),
      };
    }
  }
}

export function newUserRoleRepository(resources: Resources): IUserRoleRepository {
  return new UserRoleRepository(resources);
}
